File: Assignment 1/dilsin_amirtharajah_assignment1.cpp
#include <iostream>
#include <string>     // strings
#include <map>        // maps
#include <set>        // sets
#include <iomanip>    // decimals
using namespace std;  

int main() {
    string income;  
    double total_income = -1; // storing as double for checking

    while (true) {
        cout << "Enter monthly income: ";  // prompt one
        cin >> income;  // reading inputs

        try {
            total_income = stod(income);  // stod means string to double
            if (total_income < 0) {
                cout << "Error: Income must be non-negative.\n";
                continue; // prompt again
            }
            break; // income good
        } catch (invalid_argument const &e) {
            cout << "Error. Input must be a numerical value.\n";
        }
    }

    set<string> categories;  // storing categories
    map<string, double> expenses;  // storing expenses

    bool on = true;  // true/false 
    
    string user_prompt;  // storing category input
    string prompt = "Enter expense category or ('done'): "; // repeat

    while (on) {  // loop
        cout << prompt;   // prompt question
        cin >> user_prompt;  // get input from 

        
        if (user_prompt == "done") { // ending program
            on = false;  

            cout <<"\n---Budget Summary---\n";
            double total = 0.0;  // tracking total spent

            
            for (auto const &pair : expenses) { // looping expenses
                // auto = compiler checks type automatically
                // const =  no changing
                // & = making a copy
                total += pair.second;  // expense amount
            }
            
            double amount_left = total_income - total;  // money left
            int number_of_expenses = expenses.size();   // number of categories added

            double most_expense = 0.0;  // storing biggest expense
            for (auto const &pair : expenses) {  
                if (pair.second > most_expense) {
                    most_expense = pair.second;  // change if there is a bigger expense
                }
            }

            // decimals to two places
            cout << fixed << setprecision(2);
            cout << "Income: " << total_income << endl;
            cout << "Total spent: " << total << endl;
            cout << "Remaining: " << amount_left << endl;
            cout << "Expenses entered: " << number_of_expenses << endl;
            cout << "Largest expense: " << most_expense << "\n\n";

            cout << "By category: \n";
            // columns
            cout << left << setw(15) << "Category" << right << setw(10) << endl;
            for (auto const &pair : expenses) {
                cout << left << setw(15) << pair.first 
                     << right << setw(10) << fixed << setprecision(2) << pair.second << endl; // pair.first = key to category, pair.second = value of amount
            }
        } 
        else { // checking if input is all characters
            bool characters = true;  // if valid
            for (char c : user_prompt) {  // check each character
                if (!isalpha(c)) {  // if not a letter etc.
                    characters = false;
                    break;
                }
            }
            if (!characters) {
                cout << "Error: Input must be words\n";
                continue;  // back to start of loop
            }

            if (categories.find(user_prompt) != categories.end()) { // check if the category was used
                cout << "Error: Category must have a new name.\n";
                continue;  
            }

            string value;
            cout << "Enter amount: ";
            cin >> value;

            try {
                double amount_check = stod(value);  // change to double

                if (amount_check > 0) {
                    expenses[user_prompt] = amount_check;  // add to map
                    categories.insert(user_prompt);        // add to set
                } else if (amount_check < 0) {
                    cout << "Error. Expense amounts must be non-negative.\n";
                    continue;
                } else { // allow for 0 (incase)
                    expenses[user_prompt] = amount_check;
                    categories.insert(user_prompt);
                }
            } catch (invalid_argument const &e) {  // stod failure
                cout << "Error. Input must be a numerical value.\n";
                continue;
            }
        }
    }

    return 0;  // end 
}

File: Assignment 1/dilsin_amirtharajah_assignment1.py
income = input("Enter monthly income: ") # opening statement
prompt = ("Enter expense category or ('done'): ") # second statement
categories = set() # making a set to store
expenses = {} # making a dictionary to store
on = True
while on:
    user_prompt = input(prompt)

    if user_prompt.lower() == 'done':
        on = False

        print("---Budget Summary---")
        income_total = float(income) # making income decimals
        total = sum(expenses.values()) # adding the values of expenses
        remaining_amount = income_total - total # subtracting
        expense_amount = len(expenses) # check number of items
        big_expense = max(expenses.values()) if expenses else 0 # largest expense

        # printing to two decimal places
        print(f"Income:  {income_total: .2f}")
        print(f"Total spent:  {total: .2f}")
        print(f"Remaining:  {remaining_amount: .2f}")
        print(f"Expenses entered:  {expense_amount: .2f}")
        print(f"Largest expense:  {big_expense: .2f}\n")
        
        print("By category: ")
        for cat, amt in expenses.items():
            print(f"{cat} {amt:.2f}")

    else:
        try:
            prompt_check = user_prompt.isalpha() # checking if input is words only
            if not prompt_check:
                print("Error: Input must be a word(s).")
                continue # go back to prompt 

            if user_prompt in categories: # checking if the input was used already
                print("Error: Category must have a new name.")
                continue # go back to prompt

            value = input("Enter amount: ") # third statement
            value_check = float(value) # checking if input is numbers only
            if value_check > 0: #checking if value is greater than 0
                expenses[user_prompt] = value_check # adding value to dictionary
                categories.add(user_prompt)  # only add to categories when valid
            elif value_check < 0: # checking if value is less than 0
                print("Error. Expense amounts must be non-negative.")
                continue
            else:
                # allow 0, but add
                expenses[user_prompt] = value_check
                categories.add(user_prompt)
        except ValueError:
            print("Error. Input must be a numerical value.")
            continue # go back to prompt
